File: algorithms/S/Shortest Subarray to be Removed to Make Array Sorted/Shortest Subarray to be Removed to Make Array Sorted.java
class Solution {
    public int findLengthOfShortestSubarray(int[] arr) {
        int firstLast=0,lastFirst=arr.length-1;
        for(;firstLast<arr.length-1;firstLast++){
            if(arr[firstLast]>arr[firstLast+1]) break;
        }
		//Base case for a non-decreasing sequence
        if(firstLast==arr.length-1) return 0;
        for( ;lastFirst>0;lastFirst--){
            if(arr[lastFirst]<arr[lastFirst-1]) break;
        }
		//Possibilities 1 or 2 as mentioned above
        int minLength=Math.min(arr.length-firstLast-1,lastFirst);
        for(;firstLast>=0;firstLast--){
            for(int i=lastFirst;i<arr.length;i++){
                if(arr[firstLast]>arr[i]) continue;
                minLength = Math.min(minLength,i-firstLast-1);
                break;
            }
        }
        return minLength;
    }
}

File: scripts/algorithms/S/Shortest Subarray to be Removed to Make Array Sorted/Shortest Subarray to be Removed to Make Array Sorted.js
var findLengthOfShortestSubarray = function(arr) {
    const n = arr.length;
    
    if (n <= 1) {
        return 0;
    }
    
    let prefix = 1;
    
    while (prefix < n) {
        if (arr[prefix - 1] <= arr[prefix]) {
            prefix++;
        } else {
            break;
        }
    }
    
    if (prefix === n) {
        return 0;
    }
    
    let suffix = 1;
    
    while (suffix < n) {
        const i = n - 1 - suffix;
        
        if (arr[i] <= arr[i + 1]) {
            suffix++;
        } else {
            break;
        }
    }

    let res = Math.min(n - prefix, n - suffix);
    let left = 0;
    let right = n - suffix;

    while (left < prefix && right < n) {
        if (arr[left] <= arr[right]) {
            res = Math.min(res, right - left - 1);
            left++;
        } else {
            right++;
        }
    }

    return res;
};
